Cubical/Experiments/truncations.agda

```diff
@@ -0,0 +1,165 @@
+{-# OPTIONS --cubical --safe #-}
+module Cubical.Experiments.truncations where
+
+open import Cubical.Data.Nat.Base
+open import Cubical.Core.Prelude
+
+isOfHLevel : ∀ {ℓ} → ℕ → Set ℓ → Set ℓ
+isOfHLevel 0 A = isContr A
+isOfHLevel 1 A = isProp A
+isOfHLevel (suc (suc n)) A = (x y : A) → isOfHLevel (suc n) (x ≡ y)
+
+isGroupoid : Set → Set
+isGroupoid = isOfHLevel 3
+
+isTwoGroupoid : Set → Set
+isTwoGroupoid = isOfHLevel 4
+
+transport' : (p : I → Set) → (p i0) → (p i1)
+transport' p a = transp (λ i → p i) i0 a
+
+data propTrunc (A : Set) : Set where
+  pinc : A → propTrunc A
+  pIsProp : isProp (propTrunc A)
+
+propTruncElim : (A : Set) (B : (propTrunc A) → Set)
```

Comment:
Isn't this proof definitionally equal to https://github.com/agda/cubical/blob/master/Cubical/Core/PropositionalTruncation.agda#L35 ?

Reply:
> Isn't this proof definitionally equal to https://github.com/agda/cubical/blob/master/Cubical/Core/PropositionalTruncation.agda#L35 ?

Yeah, more or less yes. Although `refl` won't fill a path between those two in Agda for some reason. 

However I do think `propTruncElim` can have a simpler proof, namely

```Agda
propTruncElim' : {A : Set} {B : (propTrunc A) → Set} (bP : (x : propTrunc A) → isProp (B x))
  (f : (x : A) → B (pinc x)) (x : propTrunc A) → B x
propTruncElim' bP f (pinc x) = f x
propTruncElim' {B = B} bP f (pIsProp x y i) = L i
  where
  L' : PathP (λ i → B (pIsProp x y i)) (propTruncElim' bP f x) _
  L' i = transp (λ j → B (pIsProp x y (i ∧ j))) (~ i) (propTruncElim' bP f x)
  L : (i : I) → B (pIsProp x y i)
  L i = hcomp (u i) (L' i)
    where
    u : (i j : I) → Partial (i ∨ ~ i) (B (pIsProp x y i))
    u i j (i = i0) = propTruncElim' bP f x
    u i j (i = i1) = bP y (L' i1) (propTruncElim' bP f y) j
```

Note that `transp ... (~ i) ...` which simplifies a lot. The proof is basically a `compPath` except one of the paths is a dependent one. I will try to generalize this to other eliminators.

Reply:
Hmm, so this is simpler than the proof of `propTruncElim` below as the transport in B0 computes to 1 hcomp with 2 transp as sides? It's really hard to for me to beta reduce the proof below in my head though...

Reply:
It's a `hcomp` with a `transp` as base, constant on one side, and a path given by `bP` on the other side. Does this help?

I want to tell how simple a thing is by asking Agda to compute it's normal form and check the length, but often it doesn't expand things in `where` clause. Isn't it supposed to expand everything? Is this a known issue? Sometimes normal forms end up containing things like `truncations.u` (where `truncations` is the name of the module).

Reply:
I would also like this. It would be useful if it would print the number of `transp` and `hcomp` in the term as that's a good measure for how complicated something is to compute with.